### PR TITLE
Create possibility for nested subforms

### PR DESCRIPTION
**Pull request by continga**
Pull Request for Issue #11551

### Summary of Changes
This PR adds the possibility to nest subforms into subforms, adding the possibility to stack them as many times as a user would like to.

### Testing Instructions
Create a form with a subform inside of a subform and test whether it works as expected (with multiple=true or false, layout=repeatable or repeatable-table, different min/max values, etc...).

A simple testcase is to testwise edit e.g. `plugins/fields/checkboxes/params/checkboxes.xml`, adding e.g. the following after [line 27](https://github.com/joomla/joomla-cms/blob/2f181e2ec0cb1a4c81000896367ea3bf90e7f878/plugins/fields/checkboxes/params/checkboxes.xml#L27):
```xml
<field
	name="subformlevel2"
	type="subform"
	label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_LABEL"
	description="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_DESC"
	layout="joomla.form.field.subform.repeatable-table"
	icon="list"
	multiple="true"
>
	<form hidden="true" name="list_templates_modal" repeat="true">
		<field
			name="name_lvl2"
			type="text"
			label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_NAME_LABEL"
			size="30"
		/>
	</form>
</field>
```

After that has been added, navigate to Content -> Fields -> New in the Administrator Backend, select "Checkboxes" as Type, and play around with the nested subforms.

To do some additional tests, play around with the XML definition (adding 3rd level subforms, changing the multiple/non-multiple stuff, required status, etc...) and take a look whether everything works as expected.

### Expected result
Something like in the following screenshot, where I have stacked 3 subforms inside of each other with just some basic textfields:
image

### Additional information
This is a backwards-compatible change and hence does not require anything to be worried about (imho) - please correct me if I am wrong, maybe we can find a solution then.

I am aware that the main merge window of Joomla! 3.8 already has passed, so I guess this PR probably will target version 3.9 or 4.0, but I open this PR now because it is my first slightly bigger PR for Joomla, and I expect some feedback-cycles before we reach a merge-ready state.

I already have a question to begin with: I only committed the uncompressed JS file, is it my job to do the minification, or will some deploy-script take care of that?

Thanks in advance!

### CODE REVIEW

**Comment by brianteeman** [comment]
> I already have a question to begin with: I only committed the uncompressed JS file, is it my job to do the minification, or will some deploy-script take care of that?

For Joomla 3.x then yes please. You can use any minifier

**Comment by brianteeman** [comment]
@continga there are several codestyle errors that have been reported here http://213.160.72.75/joomla/joomla-cms/7879 that it would be great if you could resolve - thanks

**Comment by laoneo** [comment]
Cool feature, was waiting for it!! Does that also work with the default layout?

File: libraries/joomla/form/fields/subform.php [hunk 1]
```diff
@@ -274,6 +274,12 @@ protected function getInput()
 		$data['fieldname'] = $this->fieldname;
 		$data['groupByFieldset'] = $this->groupByFieldset;
 
+		// For each rendering process of a subform element, we want to have a
+		// separate unique subform id present to could distinguish the eventhandlers
+		// regarding adding/moving/removing rows from nested subforms from their parents.
+		static $unique_subform_id = 0;
+		$data['unique_subform_id'] = ('sr-' . $unique_subform_id++);
```
Comment (Fedik) [inline libraries/joomla/form/fields/subform.php]: name confusing, technically it is counter,
each field already have an ID `$this->id`, which is unique  
just IMHO 😉 

but seems not a bad idea to resolve "buttons" conflicts 😉

File: layouts/joomla/form/field/subform/repeatable.php [hunk 2]
```diff
@@ -39,23 +39,50 @@
...
+			);
 		endforeach;
 		?>
 		<?php if ($multiple) : ?>
 		<script type="text/subform-repeatable-template-section" class="subform-repeatable-template-section">
-		<?php echo $this->sublayout($sublayout, array('form' => $tmpl, 'basegroup' => $fieldname, 'group' => $fieldname . 'X', 'buttons' => $buttons)); ?>
+			<?php /** Do a rawurlencode to not tamper with HTML elements, especially with
+			        * nested subforms (subform in subform), this might contain a
+			        * </script> tag, which else blows up our markup. */ ?>
+			<?php echo rawurlencode(trim($this->sublayout(
```
Comment (Fedik) [inline layouts/joomla/form/field/subform/repeatable.php]: I do not like it, and I would not accept it with `rawurlencode`. why? 😉 

The problem is correct, we cannot have nested `<script>` tag.
But solution not the best.

I had idea to use `<template>` tag, wich can be nested. That would be a correct solution.
But need to search/test for `<template>` polyfill  to make it work in IE.

File: media/system/js/subform-repeatable-uncompressed.js [hunk 3]
```diff
@@ -201,6 +202,45 @@
...
+		// Create 2 strings: basename + old group, and basename + new group
+		var search = '[' + basename + '][' + group + ']',
+		    replace = '[' + basename + '][' + groupnew + ']';
+		// Does our row still contain the basename + old group? This should not happen!
+		if ($row.html().indexOf(search) !== -1) {
+			console.log('Old basename+group still existant in $row html');
+		}
+		// Recursively replace our basename + old group with basename + new group
+		// inside of nested subform template elements.
+		this.recursiveReplaceNested($row, search, replace);
```
Comment (Fedik) [inline media/system/js/subform-repeatable-uncompressed.js]: hm, I have not  tested but it looks not safe.
Each instance of `subformRepeatable` should work only with "own" fields, and should not care about nested/parent etc.

Additionally, you have fixed only the input `name`, 
but it also require to fix an input `id` and label `for`.

This is one of complex issue with nested subform 😉

File: layouts/joomla/form/field/subform/repeatable.php (hunk 2 above)
Comment (dgrammatiko) [reply layouts/joomla/form/field/subform/repeatable.php]: Big thumbs up for template, in J4 we have a polyfill already!
I think another good idea is to move **all** the field specific scripts to custom elements. The advantage is pretty obvious, we define a class one and use it as many times as we want and also less dom (slow) queries...

File: media/system/js/subform-repeatable-uncompressed.js (hunk 3 above)
Comment (Fedik) [reply media/system/js/subform-repeatable-uncompressed.js]: hm, I see, you fixed only nested template, 
but the inputs attributes of this template will be calculated depend from `data-base-name`, `data-group`, later, when subform add new row

**Comment by Fedik** [comment]
@continga thanks for doing it, it is often requested feature 😉

**Comment by laoneo** [comment]
Is this ready for testing? Because I tested it and it worked, but the minified script is missing. I really would like to get that in.

**Comment by continga** [comment]
I had some other things to do the last 2 weeks, but will continue to work on this soon to include the feedback from above and fix the CI fails

**Comment by continga** [comment]
I have now included all your feedback (I at least hope so ;-)). Would be nice if you guys now could take another look. Thanks!

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on f485db81a7b1f62b9fe6cfdbed72575e5aee8511<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on f485db81a7b1f62b9fe6cfdbed72575e5aee8511<br /><br />Tested successfully with 3 level subforms like the example<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by continga** [comment]
Is there anything we can do for merging this PR. If it's merged we like to go on with more PRs to add repeatable subforms to custom fields.

**Comment by ghost** [comment]
RTC after two successful tests.

Thanks for Hint, @continga

**Comment by tonypartridge** [comment]
@mbabker why the 3.9 release? Can we get this in 3.8.4 please since it’s a big fix as there is nothing to say that it can only be one level deep? It should be multiple and therefore a bug.

**Comment by ghost** [comment]
@tonypartridge as far as i know: new Features are shipped with new Release like 3.9., 3.8.* are for Fixes.

**Comment by tonypartridge** [comment]
Exactly, this is a bug fix.

**Comment by ghost** [comment]
ah, thought its a new Feature (Label).

**Comment by laoneo** [comment]
As much I would like to see this pr in 3.8.x, for me this is not a bug fix. Even when it is marked as bug fix, I would be hesitant to be added in a patch release just because of the amount of code changes. I would leave it for 3.9.

**Comment by tonypartridge** [comment]
There is a 3.8.4 release planned already though.

Given documentation doesn’t say it should only be 1 level deep it should be multiple levels deep and as it is not it’s therefore a bug.

**Comment by carcam** [comment]
I totally agree. It's definitely a bug fix.

**Comment by chrisdavenport** [comment]
The documentation does not state that Joomla cannot make coffee, therefore Joomla's failure to make coffee is a bug.  I suggest that someone submits a PR to fix that in 3.8.4 as I could really use more coffee (or tea).

Maintenance releases such as 3.8.4 should always aim to increase stability.  Due to the size of this change there is a chance that it may itself contain bugs and hence decrease stability.  Therefore, imho, it belongs in the next minor release (ie. 3.9.0).

**Comment by tonypartridge** [comment]
That’s just plain stupid terminoligy Chris. The documentation and function clearly allows for selections based and showon values. The fact is it DOES NOT WORK past 1 level and therefore it is a bug. But that it isn’t just documented. It’s the fact it DOES NOT WORK.

If we were adding this for the first time then yes it is a feature, we are fixing the previously implemented feature.

**Comment by brianteeman** [comment]
Even the description you wrote says its a new feature to me

>  adds the possibility to nest subforms into subforms, adding the possibility to stack them as many times as a user would like to.

**Comment by laoneo** [comment]
Just for reference, can you point to the documentation article where it is defined that the subform field should support nested subforms?

**Comment by tonypartridge** [comment]
@laoneo again it's terminology.  The nested subforms are an additional field into a form which loads more forms and you should be able to use the showon as you do everywhere else as per the showon documentation: https://docs.joomla.org/Form_field. This allows that to function, it doesn't explicitly state about subforms but the principle is the same. 

To me again it's fixing a bug in the original implementation of the subforms.

**Comment by laoneo** [comment]
But then the side effect of this pr is to fix the showon bug. It's main purpose is to add a new feature to support subforms in subforms.

**Comment by Fedik** [comment]
@tonypartridge it not a bug fix at all, it a new and very complex feature, 
with a complex logic behind, which possible still a buggy,
and as @laoneo already noted, there was no declaration that nested forms are supported

> The nested subforms are an additional field into a form which loads more forms and you should be able to use the showon as you do everywhere else as per the showon 

subform field it is not a simple field itself, 
make it nested, not only lead to code complication but also to performance loss (depends on the complexity of the form),
so I would not compare it with other regular fields

**Comment by Polm90** [comment]
I've tested the feature in the latest dev release (with two nested forms) but I found it a bit bugged...In particoular, cliccking the "more" on the external sbuform adds another copy of the innest subform instead of the one requested.
I can't add another copy of the external form.

Thank you

File: layouts/joomla/form/field/subform/repeatable-table.php [hunk 4]
```diff
@@ -67,42 +67,64 @@
...
+				<thead>
+					<tr>
+						<?php echo $table_head; ?>
+						<?php if (!empty($buttons)) : ?>
+							<th style="width:8%;">
+								<?php if (!empty($buttons['add'])) : ?>
+									<div class="btn-group">
+										<a class="btn btn-mini button btn-success group-add-<?php echo $unique_subform_id; ?>">
+											<span class="icon-plus"></span>
+										</a>
```
Comment (Quy) [inline layouts/joomla/form/field/subform/repeatable-table.php]: Add ` aria-hidden="true"`

Comment (brianteeman) [reply layouts/joomla/form/field/subform/repeatable-table.php]: You need to have something that is meaningful for non sighted users as well

**Comment by dawe78** [comment]
Since September 11th, Joomla 3.9 Beta is available for testing, so I updated to 3.9 Beta, but nested subform problem persists! It seems like this bug is still not fixed in 3.9!

**Comment by mbabker** [comment]
> It seems like this bug is still not fixed in 3.9!

Well as you can see the pull request is still open.  And based on the last few comments, it would appear that there are a couple of issues that need to be addressed.  Until all issues are addressed and the pull request is appropriately tested, this won't be merged.

**Comment by dawe78** [comment]
>  Until all issues are addressed and the pull request is appropriately tested, this won't be merged.

So this means, nested subforms will not be added to 3.9.x because its classified as feature, not bug fix, and new features won't be added to an existing version? Do we have to wait until 3.10 or 4.0?

**Comment by ghost** [comment]
@dawe78 basically a Pull request needs minimal 2 successfully Tests, which isn't given in this Case.

**Comment by tonypartridge** [comment]
@continga could you make the code style changes so we can get two tests?

**Comment by mbabker** [comment]
It wasn't just the codestyle comments, https://github.com/joomla/joomla-cms/pull/17552#issuecomment-369317804 seems to imply there's a bug with this PR too.

**Comment by gaelicwinter** [comment]
I’m going to have a crack at this tonight and see if I can make any progress, else I’ll have to code my own implementation which is not as a great a long term solution.

**Comment by continga** [comment]
I will continue to work in this PR on this weekend and check all mentioned errors, the bug report etc. I hope I can give you a testable version latest on sunday.

To be honest, for me it seemed like there was no interest from the community in this PR, why I de-prioritized working on it. I am pretty glad to hear this is not the case, so I will gladly continue on this. I will report back here latest on monday with results.

**Comment by tonypartridge** [comment]
No we absolutely want this PR :-). It's just we all have busy lives and there are many PR's. Looking forward to the update.

**Comment by dawe78** [comment]
> I will continue to work in this PR on this weekend and check all mentioned errors, the bug report etc. I hope I can give you a testable version latest on sunday.

Great, thank you so much! Let me know if I can help you, e.g testing!

**Comment by ghost** [comment]
@dawe78 if this PR is ready for Test it will be written here so you get automatically informed.

**Comment by Fedik** [comment]
for note: this already implemented for Joomla! 4 #19184

**Comment by continga** [comment]
Hi, I updated the PR to include all feedback so far. I would be happy if some of you guys could test this so I could work on possible feedback

I was not able to reproduce any issue you mentioned, @Polm90 - can you maybe be more specific about that, so what did you do, how was the result, what did you expect?

Additionally, I don't understand why the Drone Build was failing, it doesn't look to be related to any changes I made. Maybe the build had a temporary error, can I somehow restart that?

**Comment by wojsmol** [comment]
@SniperSister Please take a look on this dron failure as this is related to rips.

**Comment by ReLater** [comment]
Remove RTC if you want new tests, please. I've never looked into this pr because of the RTC label.
And it's definitely 3.10 now we have to test with? Because the pr says "wants to merge 2 commits into **joomla:staging**" (=3.9 atm)
Thaaank you!

**Comment by mbabker** [comment]
> And it's definitely 3.10 now we have to test with? Because the pr says "wants to merge 2 commits into **joomla:staging**" (=3.9 atm)

Testing against staging is fine unless this needs code that's merged only to the 3.10 branch.  Otherwise, someone can deal with rebasing branches when merging.

**Comment by SniperSister** [comment]
@wojsmol fixed

**Comment by continga** [comment]
Well, the RTC label was added after two successful tests last year in October... Afterwards, additional reviewing was done (with some feedback regarding styling/accessibility), and a rebase was needed due to conflicts. I think this is why the milestone then was changed to 3.10 in May (as those points haven't been taken care of), and it's the reason why I would like people to re-test this.

Is there any chance that we still could get this into 3.9?

If *not*, we should remove RTC label and start the whole procedure from scratch again.

If *yes*, I think we should give this high priority and merge it into staging soon, as the base-code has 2 successful tests (the base-code has not been changed since, it was only minor styling/formatting stuff), and as the PR works and has no major bugs or regressions; at least from my point of view everything works.

**Comment by mbabker** [comment]
If folks will test it out this week I've got no issue merging it (technically beta is supposed to mean feature freeze but I've always given a little flexibility to that if something is "close" (arbitrary personal measurement)).  But let's have it merged by the time beta 3 goes out next week if that's going to happen.

**Comment by tonypartridge** [comment]
Thanks @mbabker I’ll get it tested tonight   
  

  
  

  
  
>

**Comment by carcam** [comment]
I have been using this patch for months using patch tester and it has been working fine for me (not beatiful design, but with no bugs). I'll test it tomorrow with the new changes.

**Comment by ReLater** [comment]
I have tested this item :red_circle: unsuccessfully on 2bb31c0a4d2d41473e51b2d3ff87ac49528dbe76<br /><br />Reason will follow on GitHub https://github.com/joomla/joomla-cms/pull/17552<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by ReLater** [comment]
Joomla 3.9 nightly.

I installed a module that has already a subform like this in module manifest:
```
<fieldset name="addImg" label="FIELDSET_IMAGES" description="IMAGES_DESC">

	<field name="images" type="subform" min="1" max="100" multiple="true"
		formsource="modules/mod_flexsliderghsvs/myforms/subform.xml" 
		layout="joomla.form.field.subform.repeatable" groupByFieldset="false"
		label="Add images"
		description="" />

</fieldset>
```
###### simplified formsource modules/mod_flexsliderghsvs/myforms/subform.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<form>
 <field name="foto" type="media" label="Photo" description="" default="" preview="true" />
</form>
```
Module works as expected. Form looks OK.

Then I installed the patch 17552 via PatchTester. **Nothing more!!**

Afterwards my old form has 2 `Add` buttons. The bigger one is OK but doesn't work. The smaller one comes from patch and works.

**Comment by ReLater** [comment]
When I change
`layout="joomla.form.field.subform.repeatable-table"`
to
`layout="joomla.form.field.subform.repeatable"`
of example code of Testing Instructions that form has 2 buttons, too.

**Comment by Quy** [comment]
Remove RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by continga** [comment]
@ReLater Thank you for testing, bug is fixed now!

File: libraries/joomla/form/fields/subform.php [hunk 5]
```diff
@@ -274,6 +274,14 @@ protected function getInput()
 		$data['fieldname'] = $this->fieldname;
 		$data['groupByFieldset'] = $this->groupByFieldset;
 
+		/**
+		 * For each rendering process of a subform element, we want to have a
+		 * separate unique subform id present to could distinguish the eventhandlers
```
Comment (Quy) [inline libraries/joomla/form/fields/subform.php]: Remove `could`?

**Comment by tonypartridge** [comment]
I have tested this item :white_check_mark: successfully on cdbe0c92063133739c521a4cb29f02c1dfc78674<br /><br />Working well<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

I modified a module, added a form to /forms/ then a subsub form to forms/

Included and run before patch, repeating field goes crazy. 

Include and run after patch, forms are respectable, + and x toggles work as expected for each form.

Tested saving, saves as expected and reloads with the data passed into the form.

**Comment by continga** [comment]
@ReLater can you re-test this? If you give your GO we have two successful test :slightly_smiling_face: :blush:

**Comment by carcam** [comment]
I have tested this item :white_check_mark: successfully on cdbe0c92063133739c521a4cb29f02c1dfc78674<br /><br />I have successfully tested this patch in Joomla! 3.8.12 and 3.9 beta  with Patch tester.
I have successfully tested this path directly from the commit cbe0c9 code.

Test performed:

I have added this field to one of my components configuration:

```
<field
	name="subformlevel2"
	type="subform"
	label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_LABEL"
	description="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_DESC"
	layout="joomla.form.field.subform.repeatable-table"
	icon="list"
	multiple="true"
>
    <form hidden="true" name="list_templates_modal" repeat="true">
	<field
		name="name_lvl2"
		type="text"
		label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_NAME_LABEL"
		size="30"
		/>
	<field
		name="subformlevel3"
		type="subform"
		label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_LABEL"
		description="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_DESC"
		layout="joomla.form.field.subform.repeatable"
		icon="list"
		multiple="true"
	>
	<form hidden="true" name="third_level" repeat="true">
	<field
		name="name_lvl3"
		type="text"
		label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_NAME_LABEL"
		size="30"
		/>
	</form>
	</field>
    </form>
</field>
```

and in all scenarios the patch has worked with no issues.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by dawe78** [comment]
First of all, thanks for developing, testing and adding nested subforms to beta3!!!

I updated to 3.9.0 beta3 on saturday and startet testing my component using nested subform, but I got an issue... 

Steps:
1. First subform name "order_articles"
2. inside subform order_articles, nested subform "order_addons"

Click add-row for new article: row created width id "jform_order_articles__order_articles0__[nameofFields]"
Click add-row for new addon inside article: row created "jform_order_articles__order_articles1__order_addons__order_addons0__addon"

As you can see, nested subform count up article row num; subform should be jform_order_articles__order_articles0_..., but it creates jform_order_articles__order_articles1_...

I tested this behaviour only with my own component form until now, but I will do this test with a blank test form as soon as possible and report the results. Maybe someone has the same issue?

**Comment by dawe78** [comment]
Okay, tested on blank form, same issue... Nested subform starts counting on parent subform row num, eg row "order_articles2" subform starts at "order_articles3__order_addons__order_addons0".

**Comment by dawe78** [comment]
Additional information: after saving form (only parent subform rows saved) adding nested subform rows results in correct row count. So this issue occurs only on new added subform rows, after saving subform row adding nested subform rows works fine.

**Comment by zero-24** [comment]
Can you please open an new issue mention this here? As this is a closed / merged issue.

**Comment by dawe78** [comment]
Okay, sorry...

**Comment by dawe78** [comment]
Hey guys, I got an issue with nested subforms in IE 11. While nested subforms are working on FF, Chrome, Opera... , IE 11 crashed the subforms (see images below). Does anyone else got the same issue? Joomla Version is 3.9.10 (latest).

Expected result (FF):
[screen shot 2019-08-01 at 11 56 52](https://issues.joomla.org/uploads/1/de6240108535143770225316d16bf398.png)!

IE 11 Result:
[screen shot 2019-08-01 at 11 56 53](https://issues.joomla.org/uploads/1/c3becfec9cd2ebd142dd35ed231584a1.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by dawe78** [comment]
Forgot: I got this issue when I click to add a new row (main item). Existing subforms (loaded form data), including nested subforms, are displayed as expected...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by dgrammatiko** [comment]
Check the IE console, I guess the error is something like template is not defined.

Resolution ie needs a poly fill for template

**Comment by tonypartridge** [comment]
Sound like unsupported JS In IE 11. Does the console say anything?

**Comment by dawe78** [comment]
Hi dgrammatiko, thanks for your reply. 

Debugging with the console is the first thing I always do to find out whats going on. But there is no error or warning output...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by dawe78** [comment]
Hi tonypartridge, thanks too! No, unfortunally there is no error or warning output... I expected a JS warning or error, but there is none...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by dgrammatiko** [comment]
@dawe78 try adding
```html
<script src="https://unpkg.com/@webcomponents/template@1.4.0/template.js"></script>
```
at the beginning of the head section of your template (isis?). Does that fix the problem?

**Comment by dawe78** [comment]
Yes, this JS-File fixes the issue. Is there any update for ISIS available which fixes the bug? Or do I need to add your script to all my joomla installations by myself???<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by dgrammatiko** [comment]
> Is there any update for ISIS available which fixes the bug? Or do I need to add your script to all my joomla installations by myself???

The process of adding the script to the template was purely for testing purposes, the actual solution requires some deferent steps (namely calling this script before the repeatable one)

**Comment by dawe78** [comment]
Okay, is there a tutorial which describes the steps to do?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17552">issues.joomla.org/tracker/joomla-cms/17552</a>.</sub>

**Comment by dgrammatiko** [comment]
I don't think so but all is needed is adding 
```php
JHtml::_('script', 'system/template-polyfill.js', array('version' => 'auto', 'relative' => true));
```
before 
```php
	JHtml::_('script', 'system/subform-repeatable.js', array('version' => 'auto', 'relative' => true));
```
line: https://github.com/joomla/joomla-cms/blob/c372ec30c97d97f702d1dbdfd183e4d1208517c3/layouts/joomla/form/field/subform/repeatable.php#L33
Also the same in the other layout (table)

The only thing is that the js file I pointed before needs to be wrapped in an `if` statement, so it **only** executes on the browsers that they need it.
something like 
```js
  function needsTemplatePolyfill() {
    // no real <template> because no `content` property (IE and older browsers)
    var template = document.createElement('template');
    if (!('content' in template)) {
      return true;
    }
    // broken doc fragment (older Edge)
    if (!(template.content.cloneNode() instanceof DocumentFragment)) {
      return true;
    }
    // broken <template> cloning (Edge up to at least version 17)
    var template2 = document.createElement('template');
    template2.content.appendChild(document.createElement('div'));
    template.content.appendChild(template2);
    var clone = template.cloneNode(true);
    return (
      clone.content.childNodes.length === 0 ||
      clone.content.firstChild.content.childNodes.length === 0
    );
  }

if (needsTemplatePolyfill()) {

// The code of the polyfill goes here
}
```

**Comment by NathanWailes** [comment]
I'm a Joomla newbie and need to have a nested subform (subform within a subform); how do I do it?

**Comment by drmenzelit** [comment]
@NathanWailes this is a already merged pull request from 2018. Please use the documentation (docs.joomla.org or manual.joomla.org) or ask your question in the forum (forum.joomla.org).

**Pull request merged**